Let ASO use reconcile policy defined at namespace level for all its r…

antonioquintavalle1A opened the pull request:
…esources (#4600)


## What this PR does

This addressed the discussion happening in here https://github.com/Azure/azure-service-operator/issues/4600.
With this change we can specify a reconcile policy at namespace level so that all the resources will be handled according to that one, even if another value is explicitly set at resource level.

Closes #[4600]

- [x] this PR contains documentation
- [x] this PR contains tests
- [ ] this PR contains YAML Samples

matthchr commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -234,8 +235,32 @@ func (gr *GenericReconciler) createOrUpdate(ctx context.Context, log logr.Logger
+	if gr.Config.ReadReconciliationPolicyFromNamespace {
+		log.V(Debug).Info("Looking for namespace annotations")
+		// Claim resource namespace
+		namespaceObject, err := gr.KubeClient.GetObject(ctx, types.NamespacedName{Namespace: metaObj.GetNamespace(), Name: metaObj.GetNamespace()}, schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Namespace"})
```
Doing this launches a new informer for Namespaces, because we don't set `cacheOpts.ReaderFailOnMissingInformer`. I think running a namespace informer here is probably correct... but wondering if, since we're anyway paying the cost of running the namespace informer, we might as well get reconciliation triggered if namespaces are updated.

Doing this correctly requires that we set up a Watch on each resource w/ type namespace, and using the already existing `reconcilers.ARMReconcilerAnnotationChangedPredicate()`, as well as a `handler.EnqueueRequestsFromMapFunc` handler that lists the corresponding resources (in memory from cache) and triggers a reconcile on all of them.

I don't think doing this is all that difficult with the infrastructure we have set up today.

matthchr commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -351,12 +401,39 @@ func (gr *GenericReconciler) CommitUpdate(
 func (gr *GenericReconciler) handleSkipReconcile(ctx context.Context, log logr.Logger, obj genruntime.MetaObject) error {
-	reconcilePolicy := reconcilers.GetReconcilePolicy(obj, log, gr.Config.DefaultReconcilePolicy) // TODO: Pull this whole method up here
+
+	var reconcilePolicy annotations.ReconcilePolicyValue
```
It feels like this whole section where we read the reconcilePolicy is somewhat duplicated - can we factor into a shared method on `gr` that does the hierarchical lookup, and then reuse that helper function in all the locations?

matthchr replied:
Actually I think that's what `reconcilers.GetReconcilePolicy` was trying to do. Maybe delete that and make a method on `gr` that has the hierarchical lookup you want?

matthchr replied:
I think if you do that, `ParseReconcilePolicy` (that honors the default) isn't needed anymore, because that logic is encoded in the `gr` helper instead.

matthchr commented on v2/internal/reconcilers/reconcile_policy.go:
```diff
@@ -17,18 +17,27 @@ import (
+}
+
+// ParseReconcilePolicy parses the provided reconcile policy.
+// defaultPolicyValue is read from DEFAULT_RECONCILE_POLICY env variable, it set to "manage" when not specified
```
> // defaultPolicyValue is read from DEFAULT_RECONCILE_POLICY env variable, it set to "manage" when not specified

This is not accurate/required on this method.

matthchr commented on v2/internal/reconcilers/reconcile_policy.go:
```diff
@@ -17,18 +17,27 @@ import (
+	if err != nil {
+		return defaultReconcilePolicy, err
+	}
+	return reconcilePolicy, nil
```
Isn't this missing the case where the default policy exists, and the provided policy is `""`?
That shouldn't result in an error it should result in `default, nil`.

matthchr commented on v2/pkg/common/config/config.go:
```diff
@@ -101,4 +101,6 @@ const (
 	// is not explicitly defined. If omitted, it will be automatically set to "manage"
 	DefaultReconcilePolicy = "DEFAULT_RECONCILE_POLICY"
+	// ReadReconciliationPolicyFromNamespace boolean is used to determine if we need to read reconcile policy from namespace object
+	ReadReconciliationPolicyFromNamespace = "READ_RECONCILIATION_POLICY_FROM_NAMESPACE" 
```
I'm not sold we actually need an opt-in for this capability. Can you expand on your logic for why it's opt-in rather than always-on?

antonioquintavalle1A replied:
Opt-in has been provided because of the fact that it will change the expected behavior of the operator.
Today you can have 2 resources in the same namespace with a different reconcile policy.
Once activate this one, if an annotation is specified at namespace level, resources will suddenly be managed according to that policy.
This might have a side effect that might not be explictly wanted by ASO users.

theunrepentantgeek reviewed (commented):
If `READ_RECONCILIATION_POLICY_FROM_NAMESPACE` is set, this implementation looks _only_ on the namespace for a reconciliation policy, ignoring anything on the individual resource. It's effectively a switch between "resource based annotations only" and "namespace based annotations only".

I'm not sure that's the best way to achieve the desired outcome - for one, I don't think other users would expect ASO to completely ignore annotations on individual resources. 

I'd suggest turning this inside out, so to speak:

If there is no annotation on a specific resource, check the namespace to see if there is one there.

* Achieves the same goal, of allowing control at the namespace level.
* Adds the ability to override for a specific resource if desired.
* Consistent with other things in ASO (such as credentials) where a setting on an individual resource takes precedence over a wider one.
* Doesn't require any configuration.

What do you think?

antonioquintavalle1A commented:
> If `READ_RECONCILIATION_POLICY_FROM_NAMESPACE` is set, this implementation looks _only_ on the namespace for a reconciliation policy, ignoring anything on the individual resource. It's effectively a switch between "resource based annotations only" and "namespace based annotations only".
> 
> I'm not sure that's the best way to achieve the desired outcome - for one, I don't think other users would expect ASO to completely ignore annotations on individual resources.
> 
> I'd suggest turning this inside out, so to speak:
> 
> If there is no annotation on a specific resource, check the namespace to see if there is one there.
> 
>     * Achieves the same goal, of allowing control at the namespace level.
> 
>     * Adds the ability to override for a specific resource if desired.
> 
>     * Consistent with other things in ASO (such as credentials) where a setting on an individual resource takes precedence over a wider one.
> 
>     * Doesn't require any configuration.
> 
> 
> What do you think?

Rewieving this I guess that everybody agrees on the fact that resource override must be kept and namespace should not be the solely source of truth.
I'll re-adapt the code to support such use case and we'll check it again the overall change as soon as it gets ready.

theunrepentantgeek commented:
How are you going @antonioquintavalle1A? Do you need any assistance?

theunrepentantgeek commented:
The approach looks good - but currently the code does not compile.

The function `GetReconcilePolicy` returns a single result:

``` go
func GetReconcilePolicy(
        obj genruntime.MetaObject, 
        log logr.Logger, 
        defaultReconcilePolicy annotations.ReconcilePolicyValue) annotations.ReconcilePolicyValue {
```

but invocations are expecting an error as well, e.g. `generic_reconciler.go`:

``` go
reconcilePolicy, err := reconcilers.GetReconcilePolicy(obj, log, gr.Config.DefaultReconcilePolicy)
```

theunrepentantgeek commented:
> @theunrepentantgeek, effectively I missed to stage a file in the changeset. This is why it compiled locally and not here. Sorry for that.

We've all done that from time to time.

theunrepentantgeek commented:
The build of your stuff looks good - unfortunately the CI run is failing because of some broken links in the documentation. :cry:

I have a PR to fix this (#4786) - once that's completed, merge `main` into your branch and I expect the CI build will pass and @matthchr will merge this for you (I'm on leave after today).

matthchr commented on v2/internal/reconcilers/generic/register.go:
```diff
@@ -159,3 +169,30 @@ func register(
+		// list the objects for the current kind
+		for _, el := range aList.Items {
+			if _, ok := obj.GetAnnotations()["serviceoperator.azure.com/reconcile-policy"]; ok {
+				options.LogConstructor(nil).V(Status).Info("Reconciling object", el.GetName())
```
I think remove this log entirely? It will already be logged by the reconciler when it gets the event. A log outside of this loop that just says something like "detected namespace skip-reconcile annotation change" might be worth having though?

matthchr replied:
and in terms of this log, I'm fine with it at `Verbose` or `Debug` level.

matthchr replied:
Ahh yes sorry I missed the namespace (called obj) versus object (called el) nuance here on first read.

I don't strictly speaking think the "check of obj has the annotation set" is required, but it doesn't hurt so yes I'm happy with that. What I was getting at above is, I don't think we want to re-reconcile the `el` if _it_ has the annotation, because if it does, its annotation takes precedence anyway so there's no point to re-reconciling?

matthchr replied:
I'm not sure I follow your point here:
> In that case, we watch two distinct objects: ASO CR and namespace. A predicate will be aware of the change in one of these objects at the time but not for both at the same time.

You call:
```
		if err := kubeclient.List(ctx, aList, &client.ListOptions{Namespace: obj.GetName()}); err != nil {
			return []reconcile.Request{}
		}
```

If a given item in this list (say the `Foo` `ResourceGroup`), has the `serviceoperator.azure.com/reconcile-policy` set, then reconciling that resource is (AFAIK) useless. The only case I can think of where it wouldn't be useless is if actually the annotation on that object was removed, but our cache just wasn't updated yet. In this case though, we could still not emit an event here, and we will emit an event when we observe the cache update for the annotation change on the resource itself.

basically if you do:
```
for _, el := range aList.Items {
    if _, ok := el.GetAnnotations()["serviceoperator.azure.com/reconcile-policy"]; ok {
        continue // No point emitting event here as object already has its own annotation that takes precedence.
    }
```

That solves this? Am I missing the case where the above is actually not correct and we should emit an event to reconcile an object that looks to have its own `serviceoperator.azure.com/reconcile-policy` specified?

FWIW, if your argument is just "this is an optimization we don't need", that's fair enough - I'm just trying to make sure I understand if my mental model here is wrong and I'm missing something.

matthchr commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -382,3 +383,31 @@ func (gr *GenericReconciler) writeReadyConditionErrorOrDefault(ctx context.Conte
 	return err
 }
+
+func (gr *GenericReconciler) mergeReconcilePolicy(ctx context.Context, log logr.Logger, obj genruntime.MetaObject) annotations.ReconcilePolicyValue {
```
I don't love this `mergeReconcilePolicy` because it feels like it and `GetReconcilePolicy` are trying to do the same thing.

I think maybe we should just delete `reconcilers.GetReconcilePolicy` (it's only used here now) and move the logic it has to here. Also undo your changes to ParseReconcilePolicy (it should only error if it doesn't understand the policy).

Then this function becomes:
```
func (gr *GenericReconciler) mergeReconcilePolicy(ctx context.Context, log logr.Logger, obj genruntime.MetaObject) annotations.ReconcilePolicyValue {
	// We initially get the reconcile policy from the object itself
	source := "default" // assume the source is the default policy for now - this source field is used only for logging purposes
	policyStr := obj.GetAnnotations()[annotations.ReconcilePolicy]

	// If the policy is not defined at object level, then we check if it's defined at namespace level
	if policyStr == "" {
		namespaceObject, err := gr.KubeClient.GetObject(ctx, types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetNamespace()}, schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Namespace"})
		// if we cannot get the namespace, we return default reconcile policy
		if err != nil {
			log.V(Verbose).Info("Error while retrieving namespace object", "error", err)
			return gr.Config.DefaultReconcilePolicy // return default in case of error
		}
		policyStr = namespaceObject.GetAnnotations()[annotations.ReconcilePolicy]
		if policyStr != "" {
			source = "namespace"
		}
	} else {
		source = "object" // used to track where the policy was taken from for logging purposes
	}

	reconcilePolicy, err = reconcilers.ParseReconcilePolicy(policyStr, gr.Config.DefaultReconcilePolicy)
	if err != nil {
		log.Error(
			err,
			"failed to get reconcile policy. Applying default policy instead",
			"chosenPolicy", policy,
			"policyAnnotation", policyStr)
	}
	log.V(Verbose).Info("Retrieved reconcile policy", "policy", reconcilePolicy, "source", source)
	return reconcilePolicy
}
```

antonioquintavalle1A replied:
No, I was not planning to remove it.
As matter of facts, today `mergeReconcilePolicy` wraps a call to `GetReconcilePolicy`.

Shall I proceed with the method merging as you suggested before?
Don't you think we might need the same predicate for any other future scope?

matthchr replied:
I think this is done now. Resolving this convo.

matthchr commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -382,3 +385,31 @@ func (gr *GenericReconciler) writeReadyConditionErrorOrDefault(ctx context.Conte
+		// if we cannot get the namespace, we return defaul reconcile policy
+		if err != nil {
+			// reconcilePolicy here will be set to manage in case we get an error
+			log.V(Info).Info("Error while retrieving namespace object", "error", err)
```
This still applies I think (even though this was changed to Verbose, I actually think Inf owas fine but you want the `Error` helper.

matthchr commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -382,3 +385,31 @@ func (gr *GenericReconciler) writeReadyConditionErrorOrDefault(ctx context.Conte
+		reconcilePolicy, err = reconcilers.ParseReconcilePolicy(namespaceAnnotation, gr.Config.DefaultReconcilePolicy)
+	}
+
+	log.V(Info).Info("Retrieved reconcile policy", reconcilePolicy)
```
I took a crack at what this method might look like here: https://github.com/Azure/azure-service-operator/pull/4692#discussion_r2155166167

matthchr commented:
/ok-to-test sha=cb5c185

matthchr commented:
Test:
```
[produce-markdown-summary] - Test failed: TestParseReconcilePolicy
[produce-markdown-summary] === TEST OUTPUT ===
[produce-markdown-summary] === RUN   TestParseReconcilePolicy
[produce-markdown-summary] === PAUSE TestParseReconcilePolicy
[produce-markdown-summary] === CONT  TestParseReconcilePolicy
[produce-markdown-summary]     reconcile_policy_test.go:35: 
[produce-markdown-summary]         Unexpected error:
[produce-markdown-summary]             <*eris.rootError | 0xc0004cc570>: 
[produce-markdown-summary]             empty annotation
[produce-markdown-summary]             {
[produce-markdown-summary]                 global: false,
[produce-markdown-summary]                 msg: "empty annotation",
[produce-markdown-summary]                 ext: nil,
[produce-markdown-summary]                 stack: [0x162331f, 0x162395b],
[produce-markdown-summary]             }
[produce-markdown-summary]         occurred
[produce-markdown-summary] --- FAIL: TestParseReconcilePolicy (0.00s)
```

Failed in CI - PTAL.

Live tests (see errors here): https://github.com/Azure/azure-service-operator/actions/runs/15739972785/job/44362563255 - seems to have hit an issue with the webhooks:
> Message: "conversion webhook for resources.azure.com/v1api20200601storage, Kind=ResourceGroup failed: Post \"[https://azureserviceoperator-webhook-service.azureserviceoperator-system.svc:443/convert?timeout=30s\](https://azureserviceoperator-webhook-service.azureserviceoperator-system.svc/convert?timeout=30s\)": dial tcp 10.96.252.94:443: connect: connection refused",

This is probably due to the pod crashing. Best way to understand why is to run the operator in a kind cluster locally - you should be able to do this with `task controller:kind-create-with-workload-identity` - you may need to run `task provision-oidc-storage-acct` first (and I'm not sure how many people outside of MSFT have tried this so let us know if you have issues. It _should_ work but YMMV). If you struggle with this for long let me know and I can see if I can help figure out what the issue is.

antonioquintavalle1A commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -76,6 +77,8 @@ func (gr *GenericReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 
 	log := gr.LoggerFactory(metaObj).WithValues("name", req.Name, "namespace", req.Namespace)
 
+	log.V(Info).Info("Object to reconcile", metaObj)
```
Indeed, that has been removed while you were reviewing the PR.

antonioquintavalle1A commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -382,3 +385,31 @@ func (gr *GenericReconciler) writeReadyConditionErrorOrDefault(ctx context.Conte
+
+func (gr *GenericReconciler) mergeReconcilePolicy(ctx context.Context, log logr.Logger, obj genruntime.MetaObject) annotations.ReconcilePolicyValue {
+	log.V(Info).Info("Looking for namespace annotations")
+	// Claim resource namespace
```
This is a leftover, good catch

antonioquintavalle1A commented:
> Test:
> 
> ```
> [produce-markdown-summary] - Test failed: TestParseReconcilePolicy
> [produce-markdown-summary] === TEST OUTPUT ===
> [produce-markdown-summary] === RUN   TestParseReconcilePolicy
> [produce-markdown-summary] === PAUSE TestParseReconcilePolicy
> [produce-markdown-summary] === CONT  TestParseReconcilePolicy
> [produce-markdown-summary]     reconcile_policy_test.go:35: 
> [produce-markdown-summary]         Unexpected error:
> [produce-markdown-summary]             <*eris.rootError | 0xc0004cc570>: 
> [produce-markdown-summary]             empty annotation
> [produce-markdown-summary]             {
> [produce-markdown-summary]                 global: false,
> [produce-markdown-summary]                 msg: "empty annotation",
> [produce-markdown-summary]                 ext: nil,
> [produce-markdown-summary]                 stack: [0x162331f, 0x162395b],
> [produce-markdown-summary]             }
> [produce-markdown-summary]         occurred
> [produce-markdown-summary] --- FAIL: TestParseReconcilePolicy (0.00s)
> ```
> 
> Failed in CI - PTAL.
> 
> Live tests (see errors here): https://github.com/Azure/azure-service-operator/actions/runs/15739972785/job/44362563255 - seems to have hit an issue with the webhooks:
> 
> > Message: "conversion webhook for resources.azure.com/v1api20200601storage, Kind=ResourceGroup failed: Post "[https://azureserviceoperator-webhook-service.azureserviceoperator-system.svc:443/convert?timeout=30s](https://azureserviceoperator-webhook-service.azureserviceoperator-system.svc/convert?timeout=30s)": dial tcp 10.96.252.94:443: connect: connection refused",
> 
> This is probably due to the pod crashing. Best way to understand why is to run the operator in a kind cluster locally - you should be able to do this with `task controller:kind-create-with-workload-identity` - you may need to run `task provision-oidc-storage-acct` first (and I'm not sure how many people outside of MSFT have tried this so let us know if you have issues. It _should_ work but YMMV). If you struggle with this for long let me know and I can see if I can help figure out what the issue is.

That's strange, I recompiled the operator and loaded it on an openshift cluster I have. Didn't observe any particular crash or instabilities.

Nevertheless, I'll try it on kind to see if I can replicate the issue spot by the CI.

antonioquintavalle1A commented:
Rebased on main to include https://github.com/Azure/azure-service-operator/pull/4786

matthchr commented on v2/internal/reconcilers/generic/register.go:
```diff
@@ -148,6 +156,8 @@ func register(
 		WithOptions(options.Options)
 	builder.Named(info.Name)
 
+	builder = builder.Watches(&corev1.Namespace{}, handler.EnqueueRequestsFromMapFunc(registerNamespaceWatcher(options, kubeClient, info.Obj, gvk)), ctrlbuilder.WithPredicates(predicate.AnnotationChangedPredicate{}))
```
Please change this to use internal/reconcilers/predicates.go `ARMReconcilerAnnotationChangedPredicate`

That will filter the events to only the annotation we care about.

matthchr commented on v2/internal/reconcilers/generic/generic_reconciler.go:
```diff
@@ -382,3 +383,30 @@ func (gr *GenericReconciler) writeReadyConditionErrorOrDefault(ctx context.Conte
+			log.V(Verbose).Info("Error while retrieving namespace object", "error", err)
+			return reconcilePolicy
+		}
+		log.V(Verbose).Info("Found Namespace annotations", namespaceObject.GetAnnotations()[annotations.ReconcilePolicy])
```
This should probably be log.V(Verbose).Error instead.

matthchr commented:
/ok-to-test sha=304c6d0

matthchr commented:
Let me requeue the tests. If they fail, I can take a look on Monday (it's a holiday in the US today) to help move this forward.

Thanks for your perseverance!

matthchr commented:
/ok-to-test sha=983b1f7

matthchr commented:
Ok I checked this out and there are two issues:
**Failing unit test.** 
You still need to fix this test: https://github.com/Azure/azure-service-operator/pull/4692#issuecomment-2987335164. If you do as I suggested [here](https://github.com/Azure/azure-service-operator/pull/4692/files#r2155166167) and drop `GetReconcilePolicy` entirely, you don't need `ParseReconcilePolicy` to return an error and so you can change the impl of `ParseReconcilePolicy` back and the test will pass.

**Lack of namespaces permission**:
The pod is cycling in kind because we don't have permission to read namespaces.

```
E0708 00:22:52.713545       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
E0708 00:22:55.475682       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
E0708 00:23:00.855021       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
E0708 00:23:10.846384       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
E0708 00:23:30.293361       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
E0708 00:24:00.946169       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
E0708 00:24:33.750907       1 runtime.go:226] "Failed to watch" err="failed to list *v1.Namespace: namespaces is forbidden: User \"system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope" logger="controller-runtime.cache.UnhandledError" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285" type="*v1.Namespace"
```

We need to add the namespaces read permission to the operator.  You need to add it here. https://github.com/Azure/azure-service-operator/blob/15efd612c9d2c4a2c94634ba6b1b4dfe26034788/v2/api/resources/v1api20200601/permissions.go#L4

Should be just get/list/watch - no update/patch/delete for namespaces.

matthchr commented:
> One simple way to address this, would be to give a clusterRoleBinding to allow reading of each and every namespace.

I think we need to add the permission to the set of roles the operator has, similar to how it can read secrets and configmaps. I'm not too worried about failing safe if we don't have the perms, at least for now, because as far as I know, most/all users are using the default permissions we grant anyway, and if users were going to have problems with certain permissions I would imagine it would be the configmap/secrets read, rather than namespaces read

matthchr commented on v2/internal/reconcilers/generic/register.go:
```diff
@@ -143,19 +152,57 @@ func register(
+
+// This registers a watcher on the namespace for the watched resource
+func registerNamespaceWatcher(options Options, kubeclient kubeclient.Client, object client.Object, gvk schema.GroupVersionKind) func(ctx context.Context, obj client.Object) []reconcile.Request {
+	options.LogConstructor(nil).V(Status).Info("Handler for resource object", "name", object.GetName(), "object.namespace", object.GetNamespace(), "object.kind", gvk.Kind)
```
minor: ping on this item, I think we should improve this log message.

matthchr commented on v2/internal/reconcilers/generic/register.go:
```diff
@@ -159,3 +169,30 @@ func register(
+		// list the objects for the current kind
+		for _, el := range aList.Items {
+			if _, ok := obj.GetAnnotations()["serviceoperator.azure.com/reconcile-policy"]; ok {
+				options.LogConstructor(nil).V(Status).Info("Reconciling object", el.GetName())
```
Reduce from `Status` to a more verbose log level.

matthchr reviewed (commented):
This overall LGTM, I just have a few nits about logs.

If you can clean those up and response to the latest comment on this thread, I think this is good to merge: https://github.com/Azure/azure-service-operator/pull/4692/files#r2155134742

Thanks for the perseverance.

matthchr commented:
/ok-to-test sha=cbb4b97

antonioquintavalle1A commented:
Hello there @matthchr,
I ran `controller:format-code` locally and I've got no diffs in the code.
If I understand it correctly, the error comes from the fact that a file is getting modified when it should have not, as checked here:

`  verify-no-changes:
    desc: Checks that there are no uncommitted modifications to files
    cmds:
      - cmd: 'if [ -n "{{.CHANGED}}" ]; then echo "Error: files were modified during the build (ensure you have commited generated files, and run ''task format-code'' to tidy):"; echo "{{.CHANGED}}"; git diff; exit 1; fi'
        silent: true
    vars:
      CHANGED:
        sh: git status --porcelain`

However the change here alters solely the order of declarations not the file content:

` [verify-no-changes] --- a/v2/internal/reconcilers/generic/register.go
[verify-no-changes] +++ b/v2/internal/reconcilers/generic/register.go
[verify-no-changes] @@ -10,7 +10,6 @@ import (
[verify-no-changes]  	"fmt"
[verify-no-changes]  
[verify-no-changes]  	. "github.com/Azure/azure-service-operator/v2/internal/logging"
[verify-no-changes] -	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
[verify-no-changes]  
[verify-no-changes]  	"github.com/go-logr/logr"
[verify-no-changes]  	"github.com/rotisserie/eris"
[verify-no-changes] @@ -30,6 +29,7 @@ import (
[verify-no-changes]  	"sigs.k8s.io/controller-runtime/pkg/reconcile"
[verify-no-changes]  
[verify-no-changes]  	"github.com/Azure/azure-service-operator/v2/internal/config"
[verify-no-changes] +	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
[verify-no-changes]  	"github.com/Azure/azure-service-operator/v2/internal/util/interval"
[verify-no-changes]  	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
[verify-no-changes]  	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"`

Shall we relax a bit the way local files are checked?
